src/iterator.hpp

```diff
@@ -0,0 +1,127 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#pragma once
+
+#include <functional>
+#include <set>
+#include <stdint.h>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include <IPWRuleProcessor.h>
+#include <PWManifest.h>
+#include <utils.h>
+#include <validator.hpp>
+
+// Eventually object will be a class rather than a namespace
+namespace ddwaf::object
+{
+
+class iterator_base
+{
+public:
+    explicit iterator_base(const object_limits &limits = object_limits());
+    virtual ~iterator_base() = default;
+
+    iterator_base(const iterator_base&) = default;
+    iterator_base(iterator_base&&) = default;
```

Comment:
I don't recall why I deleted the assignment, I will need to double check.

The use case for move might not be there, but at some point I plan to add an iterator cache so that we don't have to incur the cost of iterator initialisation more than once, since the iterator is trivially copyable.